---
title: "Differential Gene Expression Analysis"
subtitle: "Mycobacterium ***abscessus***"
author: "Ranjitha"
output:
  html_document:
      theme: united
      version: 4
---

```{r setup}
knitr::opts_chunk$set(echo = TRUE, tidy=TRUE, tidy.opts=list(width.cutoff=80))
```

```{r, echo=TRUE, results="hide", eval=FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("DESeq2")

BiocManager::install("ComplexHeatmap")

install.packages("tidyverse")
install.packages("ggrepel")
install.packages("FactoMineR")
install.packages('cluster')
install.packages('factoextra')
install.packages("ashr")
install.packages("circlize")
install.packages("formatR")

BiocManager::install("clusterProfiler")
BiocManager::install("enrichplot")
install.packages("data.table")```

## Required Libraries

```{r, echo=TRUE, message=FALSE, warning=FALSE, results="hide"}
library(DESeq2)
library(tidyverse)
library(ComplexHeatmap)
library(ggplot2)
library(ggrepel)
library(cluster)
library(ashr)
library(formatR)
library(EnhancedVolcano)
library(clusterProfiler)
library(enrichplot)
library(data.table)
```

## Read Input

```{r, echo=TRUE}
samples <- read.csv("counts.csv", header=T, row.names=1, sep=",",check.names = FALSE)
proteinkey <- read.delim("protein.txt", header = FALSE)
KOkey <- read.delim("ID_KO.txt", header = FALSE)
col <- read.table("condition.txt", header=T, row.names=1, sep="\t", stringsAsFactors = TRUE)
```

## Data Cross-verification

```{r, echo=TRUE, results='hold'}
all(colnames(samples) %in% rownames(col))
all(colnames(samples) == rownames(col))
```

## Set Reference Condition

```{r, echo=TRUE, results='hide'}
col$Condition <- relevel(col$Condition, ref="Control") 
model.matrix(~Condition,data=col)
```

## Create Design Matrix

```{r, echo=TRUE, results='hide'}
dds <- DESeqDataSetFromMatrix(samples,col,~Condition)
keep <- rowSums(counts(dds)) >= 0 #Discard low counts
dds <- dds[keep,]
dds
```
```{r}
dds <- estimateSizeFactors(dds)
dds <- estimateDispersions(dds)
plotDispEsts(dds)
```


## PCA

```{r}
w <- rlog(dds, blind=TRUE)
png("PCA_Scatter.png", res=100, width =700 , height =600)

z <- plotPCA(w,intgroup=c("Condition"))+ stat_ellipse()
z + ggtitle("Batch Effect Verification of Replicates") + geom_label_repel(aes(label = name), size=8) + theme(legend.position = "None", panel.background = element_rect(fill = '#eaf1f8'), plot.margin = unit(c(2, 2, 1, 1), "cm"))
```

## Run DESeq2

```{r, echo=TRUE, message=FALSE, warning=FALSE}
ddsDE <- DESeq(dds)
resultsNames(ddsDE)
```


## LogFoldChange Contrasts

```{r, echo=TRUE}
Condition_Erythromycin_vs_Control <- results(ddsDE, contrast=c("Condition",  "Erythromycin", "Control"), independentFiltering=TRUE, alpha=0.05, pAdjustMethod="BH", parallel=TRUE)

Condition_Kanamycin_vs_Control <- results(ddsDE, contrast=c("Condition", "Kanamycin", "Control"), independentFiltering=TRUE, alpha=0.05, pAdjustMethod="BH", parallel=TRUE)
```

## LogFoldChange Shrink

```{r, echo=TRUE, message=FALSE, warning=FALSE}
Condition_Erythromycin_vs_Control <- lfcShrink(ddsDE, contrast=c("Condition",  "Erythromycin", "Control"), res=Condition_Erythromycin_vs_Control, type="ashr")

Condition_Kanamycin_vs_Control <- lfcShrink(ddsDE, contrast=c("Condition",  "Kanamycin", "Control"), res=Condition_Kanamycin_vs_Control, type="ashr")
```

## Control Selection

### Create gene name key

```{r, echo=TRUE, results='hide', warning=FALSE, message=FALSE}
proteinkey_df <- as.data.frame(proteinkey)
samples$Gene_Name <- proteinkey_df$V1
samples$KO_annot <- KOkey$V2
#write.table(rownames(samples),"samples.csv", sep=",")
```


## Create Data Frames and Appending Gene Name

```{r, echo=TRUE, results='hide'}
Condition_Erythromycin_vs_Control_df <- as.data.frame(Condition_Erythromycin_vs_Control)
Condition_Erythromycin_vs_Control_df$Gene_Name <- proteinkey_df$V1
Condition_Erythromycin_vs_Control_df$KO_annot <- KOkey$V2

Condition_Kanamycin_vs_Control_df <- as.data.frame(Condition_Kanamycin_vs_Control)
Condition_Kanamycin_vs_Control_df$Gene_Name <- proteinkey_df$V1
Condition_Kanamycin_vs_Control_df$KO_annot <- KOkey$V2
```

## Data Sorting and tidying

```{r, echo=TRUE, results='hide'}
Ordered_Erythromycin_vs_Control <- Condition_Erythromycin_vs_Control_df[order(Condition_Erythromycin_vs_Control_df$padj,-Condition_Erythromycin_vs_Control_df$log2FoldChange),]
Ordered_Erythromycin_vs_Control <- na.omit(Ordered_Erythromycin_vs_Control)

Ordered_Kanamycin_vs_Control <- Condition_Kanamycin_vs_Control_df[order(Condition_Kanamycin_vs_Control_df$padj,-Condition_Kanamycin_vs_Control_df$log2FoldChange),]
Ordered_Kanamycin_vs_Control <- na.omit(Ordered_Kanamycin_vs_Control)
```

## Pathway Analysis

```{r, echo=TRUE, results='hide'}
KOkey$V2 <- gsub( "ko:", "", as.character(KOkey$V2))
kegg <- data.table(KOkey)
kegg <- kegg[, list(V2 = unlist(strsplit(V2 , ","))), by = V1] 
kegg_final <- kegg[,c(2,1)]

protein_ids <- rownames(Ordered_Erythromycin_vs_Control)
protein_ids <- as.data.frame(protein_ids)

keep <- subset(kegg_final,protein_ids  %in% protein_ids)

enr_results <- enrichKEGG(keep$V2, organism='ko', minGSSize = 1, pvalueCutoff = 0.05, qvalueCutoff = 0.05)
png("KO_dotplot.png", width = 500, height = 900)
dotplot(enr_results, showCategory=20)
dev.off()
```
# Manually selecting/checking for genes, Modified according to substrate selection

```{r, echo=TRUE, results='hide', tidy=TRUE, tidy.opts=list(width.cutoff=80)}

check_genes<-Ordered_Erythromycin_vs_Control[grepl("High-affinity branched-chain amino acid transport|Aminoglycoside 2'-N-acetyltransferase|putative multidrug-efflux transporter",Ordered_Erythromycin_vs_Control$Gene_Name),]

```

## Volcano Plots
### Modified volcano according to substrate selection

```{r, fig.height=7, echo=TRUE, warning=FALSE, message=FALSE, results='hide'}

png("Ordered_Kanamycin_vs_Control.png", width = 1000, height = 900)
  EnhancedVolcano(Ordered_Kanamycin_vs_Control,
    lab = Ordered_Kanamycin_vs_Control$Gene_Name,
    title = "Kanamycin_vs_Control",
    x = 'log2FoldChange',
    y = 'pvalue',
    pCutoff =0.05,
    FCcutoff = 1,
    pointSize = 3.0,
    labSize = 6.0,
    drawConnectors = TRUE,
    widthConnectors = 0.75,
    boxedLabels = TRUE,
    colConnectors = 'grey',
    col = c("#94afbb","#e78f8e","#a6c47c", "#c65390"),
    colAlpha = 1,
    gridlines.major = FALSE,
    gridlines.minor = FALSE,
    typeConnectors="open",
    cutoffLineCol="grey") 
dev.off()
```

## Heatmap

### Assign Threshold

```{r, echo=TRUE}
log2cutoff <- 3
qvaluecutoff <- 0.05

sigGenes <- unique(c(rownames(subset(Condition_Erythromycin_vs_Control, padj<=qvaluecutoff & abs(log2FoldChange)>=log2cutoff)), rownames(subset(Condition_Kanamycin_vs_Control, padj<=qvaluecutoff & abs(log2FoldChange)>=log2cutoff))))

```



### Concatenate Geneid with Variance Stabilized DESeq Normalized Values

```{r, echo=TRUE}
v <- rlog(ddsDE)
heat <- assay(v)[sigGenes,]

```

### Generate & Modify heatmap according to gene selection

```{r, echo=TRUE, results='hide', tidy=TRUE, tidy.opts=list(width.cutoff=60)}
png("Heatmap.png", res=200, width =700 , height =850) #check here
col_fun = colorRamp2(c(5,10,15), c("#a2c8c8","white", "#ff5a5a"))
col_fun(seq(3,-3))
ht1 <- Heatmap(heat, 
               row_title = "Genes", 
               row_title_gp=gpar(fontface="bold", fontsize=8),
               column_title = "MAB (DGE LFC=3; padj=0.05)", 
               column_title_gp = gpar(fontface="bold", fontsize=8), 
               column_labels=colnames(heat), 
               row_labels=rownames(heat), 
               col=col_fun, 
               column_dend_height = unit(2,"cm"), 
               row_dend_width = unit(2,"cm"), 
               heatmap_legend_param = list(title="Gene Expression Levels",
               title_gp = gpar(fontface="bold",fontsize=8),
               at = c(5, 10, 15),
               labels = c("Low", "Medium", "High"), 
               labels_gp = gpar(fontsize=8),
               direction = "horizontal", 
               title_position = "topcenter"), 
               clustering_distance_rows = "euclidean", 
               column_km=2, 
               column_names_rot = 45,
               column_names_gp = gpar(fontsize=7),
               row_names_gp = gpar(fontsize=7),
               row_names_max_width = unit(15.2, "cm"),
               width = ncol(heat)*unit(4, "mm"), 
               height = nrow(heat)*unit(3, "mm"))
draw(ht1, heatmap_legend_side = "bottom")
```

